### PR TITLE
[ADD] delivery_provider: Shipping providers interface

### PR DESCRIPTION
**Pull request by rim-odoo**
Basic api for shipping provider + demo provider

### CODE REVIEW

**Comment by rim-odoo** [comment]
@tde-banana-odoo @antonylesuisse Ready for your comments !

File: addons/delivery_provider/models/delivery_carrier.py [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+            else:
+                res = super(DeliveryCarrier, self).get_price('price', [])
+                self.available = res[self.id]['available']
+                self.price = res[self.id]['price']
+
+    #########################################
+    # Hooks to implement external providers #
+    #########################################
+
+    # TODO definir et gerer les exceptions qui peuvent etre lancees par ces providers
```
Comment (KangOl) [inline addons/delivery_provider/models/delivery_carrier.py]: at leat in english

File: addons/delivery_provider/models/delivery_carrier.py [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+
+    def get_shipping_price_from_so(self, orders):
```
Comment (KangOl) [inline addons/delivery_provider/models/delivery_carrier.py]: add `@one` decorator

File: addons/delivery_provider/models/delivery_carrier.py [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+            # For demo: random tracking number, and price can be the estimation of the SO or 1.0 more
+            shipment = {'tracking_number': picking.id + random.randint(0, 123456789),
+                        'exact_price': 10.0 * len(picking.move_lines) + 5 + random.randint(0, 1)}
+            res = res + [shipment]
+        return res
+
+    def dummy_get_tracking_link(self, pickings):
+        res = []
+        for picking in pickings:
+            res = res + [('https://www.mydummytracker.com/track?shipid=%s' % picking[0].carrier_tracking_ref)]
```
Comment (KangOl) [inline addons/delivery_provider/models/delivery_carrier.py]: Use an existing url.
You can use its app page (will works once merged): https://www.odoo.com/apps/modules/online/delivery_provider/

File: addons/delivery_provider/models/delivery_carrier.py (hunk 2 above)
Comment (rim-odoo) [inline addons/delivery_provider/models/delivery_carrier.py]: @KangOl Why that? We could ask FedEx for the price of more than one shipping?

File: addons/delivery_provider/models/delivery_carrier.py (hunk 1 above)
Comment (rim-odoo) [inline addons/delivery_provider/models/delivery_carrier.py]: done ;-)

File: addons/delivery_provider/models/delivery_carrier.py (hunk 3 above)
Comment (rim-odoo) [inline addons/delivery_provider/models/delivery_carrier.py]: Done, but I don't think @antonylesuisse will merge this demo provider anyway (I plan to move it to a separate commit to ease the merge)

File: addons/delivery_provider/models/delivery_carrier.py (hunk 2 above)
Comment (KangOl) [inline addons/delivery_provider/models/delivery_carrier.py]: it's one carrier, not one order

Comment (rim-odoo) [inline addons/delivery_provider/models/delivery_carrier.py]: aaah yeah, you're right!

so I also should add `@api.one` to `send_shipping`, `get_tracking_link` and `cancel_shipment`?

Comment (KangOl) [inline addons/delivery_provider/models/delivery_carrier.py]: exactly. And also `<provider>_<func>` methods

Comment (rim-odoo) [inline addons/delivery_provider/models/delivery_carrier.py]: ok, let's do this!

Comment (rim-odoo) [inline addons/delivery_provider/models/delivery_carrier.py]: @KangOl But doing this will encapsulate every result in a list containing a list of results... I'm not sure, but in fact, these methods are more multi (and by the way they are already called on singletons)... I'd suggest a self.ensure_one to avoid ugly unpackings

Comment (KangOl) [inline addons/delivery_provider/models/delivery_carrier.py]: ok, let's do this!

Comment (rim-odoo) [inline addons/delivery_provider/models/delivery_carrier.py]: done

**Comment by rim-odoo** [comment]
@KangOl @antonylesuisse @tde-banana-odoo It should be ready for a merge; I moved my demo provider in another commit, so not merging it should be easy

**Comment by rim-odoo** [comment]
Merged at https://github.com/odoo/odoo/commit/3098ab7fd101b1bc5caa7ad8fd3200c0b5e66310
